### PR TITLE
Added option to make windows borderless and rounded (Cleaner)

### PR DESCRIPTION
**Pull request by DrabWeb**
image

This is a redo of the other pull request to add the same feature (https://github.com/gnachman/iTerm2/pull/233). This is much cleaner and doesn't add unnecessary changes.

### CODE REVIEW

File: sources/PseudoTerminal.m [hunk 1]
```diff
@@ -326,11 +326,13 @@ + (NSInteger)styleMaskForWindowType:(iTermWindowType)windowType {
...
         case WINDOW_TYPE_LEFT_PARTIAL:
         case WINDOW_TYPE_RIGHT_PARTIAL:
-        case WINDOW_TYPE_NO_TITLE_BAR:
             return NSBorderlessWindowMask | NSResizableWindowMask;
 
         case WINDOW_TYPE_TRADITIONAL_FULL_SCREEN:
             return NSBorderlessWindowMask;
+            
+        case WINDOW_TYPE_NO_TITLE_BAR_ROUNDED:
+            return NSTitledWindowMask | NSFullSizeContentViewWindowMask | NSClosableWindowMask | NSMiniaturizableWindowMask | NSResizableWindowMask;
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Does it have to have closable and miniaturizable? Those buttons aren't visible so it's strange that it's set.

File: sources/PseudoTerminal.m [hunk 2]
```diff
@@ -592,6 +593,32 @@ - (void)finishInitializationWithSmartLayout:(BOOL)smartLayout
 
     [self updateTabBarStyle];
     self.window.delegate = self;
+    
+    if(windowType == WINDOW_TYPE_NO_TITLE_BAR_ROUNDED) {
+        // Hide the titlebar
+        self.window.titlebarAppearsTransparent = true;
+        
+        // Hide the titlebar
+        self.window.titlebarAppearsTransparent = true;
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: This is a duplicate of line 599

File: sources/PseudoTerminal.m [hunk 3]
```diff
@@ -592,6 +593,32 @@ - (void)finishInitializationWithSmartLayout:(BOOL)smartLayout
...
+        
+        // Hide the title
+        self.window.titleVisibility = NSWindowTitleHidden;
+        
+        // Hide the close button
+        NSButton* closeButton = [[self window] standardWindowButton:NSWindowCloseButton];;
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Please put space before the *
There are two semicolons

Same on lines 612 and 616

File: sources/PseudoTerminal.m [hunk 4]
```diff
@@ -843,7 +870,6 @@ - (void)tmuxFontDidChange:(NSNotification *)notification
     // create a new terminal window
     int newWindowType;
     switch (windowType_) {
-        case WINDOW_TYPE_NO_TITLE_BAR:
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Why the changes to this method?

File: sources/PseudoTerminal.m [hunk 5]
```diff
@@ -1567,7 +1592,6 @@ + (void)drawArrangementPreview:(NSDictionary*)terminalArrangement
             rect = virtualScreenFrame;
             break;
 
-        case WINDOW_TYPE_NO_TITLE_BAR:
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Why the changes to this method?

File: sources/PseudoTerminal.m [hunk 6]
```diff
@@ -4109,8 +4131,7 @@ - (PSMTabBarControl *)tabView:(NSTabView *)aTabView
 
     NSWindowController<iTermWindowController> * term =
         [self terminalDraggedFromAnotherWindowAtPoint:point];
-    if (([term windowType] == WINDOW_TYPE_NORMAL ||
-         [term windowType] == WINDOW_TYPE_NO_TITLE_BAR) &&
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: I think this is still needed

File: sources/PseudoTerminal.m [hunk 7]
```diff
@@ -5821,7 +5842,6 @@ - (BOOL)_haveTopBorder
     BOOL topTabBar = ([iTermPreferences intForKey:kPreferenceKeyTabPosition] == PSMTab_TopTab);
     BOOL visibleTopTabBar = (tabBarVisible && topTabBar);
     BOOL windowTypeCompatibleWithTopBorder = (windowType_ == WINDOW_TYPE_BOTTOM ||
-                                              windowType_ == WINDOW_TYPE_NO_TITLE_BAR ||
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: No-title-bar windows should draw a top border

File: sources/PseudoTerminal.m [hunk 8]
```diff
@@ -6124,7 +6144,7 @@ - (void)safelySetSessionSize:(PTYSession*)aSession rows:(int)rows columns:(int)c
     }
     PtyLog(@"safelySetSessionSize");
     BOOL hasScrollbar = [self scrollbarShouldBeVisible];
-    if (windowType_ == WINDOW_TYPE_NORMAL || windowType_ == WINDOW_TYPE_NO_TITLE_BAR) {
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Did you mean to remove this? It would prevent term-initiated resizes from working for no-titlebar windows.

**Comment by DrabWeb** [comment]
@gnachman Well this is alot to go through, but I will try my best. I also have very little experience with writing Objective C, as I do my OSX apps in Swift.

For bar 333 - 335: This was necessary when I did this style windows in other apps. I can remove the close mask because it still closes without this. It needs the miniaturizable window mask so you can hit ⌘M and minimize the window.

For 596 - 602: Fixed, no longer duplicated.

For 596 - 608: Added spaces and removed the duplicate semicolons on the necessary lines.

For 846: I changed it to better represent what it is doing. I can change it back if you like.

For 4112 - 4113: Added it back.

For 5824: Added it back.

For 6127: I actually had no idea what it did, I should probably stop fiddling with these things. Fixed

File: sources/PseudoTerminal.m [hunk 9]
```diff
@@ -326,11 +326,13 @@ + (NSInteger)styleMaskForWindowType:(iTermWindowType)windowType {
...
         case WINDOW_TYPE_LEFT_PARTIAL:
         case WINDOW_TYPE_RIGHT_PARTIAL:
-        case WINDOW_TYPE_NO_TITLE_BAR:
             return NSBorderlessWindowMask | NSResizableWindowMask;
 
         case WINDOW_TYPE_TRADITIONAL_FULL_SCREEN:
             return NSBorderlessWindowMask;
+            
+        case WINDOW_TYPE_NO_TITLE_BAR_ROUNDED:
+            return NSTitledWindowMask | NSFullSizeContentViewWindowMask | NSMiniaturizableWindowMask | NSResizableWindowMask;
```
Comment (gnachman) [inline sources/PseudoTerminal.m]: Yech, I hadn't noticed that NSFullSizeContentViewWindowMask was new to 10.10. We'll have to make this conditional on OS version. I'm going to put this PR on hold for the time being because El Capitan is going to make my life a little crazy and I don't want to introduce another variable--but I do promise to come back to it.

**Comment by aljex** [comment]
This is pretty cool.

**Comment by RomainLanz** [comment]
Any news on it?

**Comment by DrabWeb** [comment]
@RomainLanz I have abandoned this pull request, as I dont feel it needs to be in the official builds. You can still compile my branch at https://github.com/DrabWeb/iTerm2 and get these features though

**Comment by RomainLanz** [comment]
It's so better if we could have a little checkbox to check in iTerm2 Settings. Do wanna compile my own version and can't update later.

**Comment by Phrohdoh** [comment]
I agree that this would be a great feature to have.
If a maintainer is interested in this I'd be willing to try taking it on (based on this PR of course).

**Comment by base-zero** [comment]
👍  Would be very nice to have this as an option in version 3 !

**Comment by jonaswouters** [comment]
@Phrohdoh I think the maintainer showed interest by commenting on the original pull request. Perhaps open a ticket on gitlab and continue from there?
